### DISCUSSION TITLE
How to create, and use, a prebuilt package

### DISCUSSION THREAD

**Discussion by perllaghu** [jupyterlab/jupyterlab#16015] [Q&A]
I have created an extension for jupyterlab 4 [using `copier`, the [extension-examples](https://github.com/jupyterlab/extension-examples), and the [develop extensions](https://jupyterlab.readthedocs.io/en/stable/extension/extension_dev.html#develop-extensions)) documentation.]

It works just lovely when I test it from within the extension folder..... but I can _not_ figure out how to install it into a dockerised jupyter image.

I have zero desire to publish an in-house, bespoke, targetted extension in a public repository like pypi.org

I have two questions:

1. How to I install an extension from git (using `RUN pip install git+https://github.com/perllaghu/noteable-laf`, the code appears in the image:
```
(base) jovyan@895df8069705:~$ find / -name noteable-laf 2>/dev/null
/opt/conda/lib/python3.11/site-packages/noteable-laf
/opt/conda/share/jupyter/labextensions/@noteable/noteable-laf
(base) jovyan@895df8069705:~$ 
``` 
2. If I need to do some magic to convert a working local extension into a distribution, what do I need to do?
(I've followed the instructions that `copier` puts into the RELEASE.md file... up to the 

> Then to upload the package to PyPI, do:
> 
> ```bash
> twine upload dist/*
> ```

.... and yes, I've put a copy of my code on github, so the actual code can be seen

**Comment by perllaghu** [comment]
Further - here's a Dockerfile that works:

```
FROM jupyter/minimal-notebook:848a82674792

WORKDIR /src

RUN git clone git+https://github.com/perllaghu/noteable-laf
WORKDIR /src/noteable-laf

RUN pip install -e "."  \
    && npm install \
    && jlpm install \
    && jlpm build \
    && jupyter labextension develop . --overwrite
WORKDIR $HOME
```

..... that's - _messy and 'orrible_

**Comment by krassowski** [comment]
I do not understand your question. If you do not want to upload the artifacts from `dist/` to PyPI you can still store them someplace else. Since you already have a GitHub repository, creating a GitHub release and uploading the tarball and a wheel file there would sound reasonable to me. Then you just fetch that tarball/wheel in your docker file and install it using `pip`.

If your extension is frontend-only (does not contain server component) you could just drop the compiled (prebuilt) assets into appropriate `share/jupyter/labextensions/` path.

> How to I install an extension from git (using RUN pip install git+https://github.com/perllaghu/noteable-laf, the code appears in the image:

Where is the question here? Does it appear in the right path but the extension does not load up?

> If I need to do some magic to convert a working local extension into a distribution, what do I need to do?

Are you asking how to build the extension to get PyPI tarball/wheel files? You should be able to do that by running `python -m build` in the extension directory, assuming that you have build dependencies (such as `build`, `hatchling` etc) installed.

**Reply by perllaghu** [reply]
Thanks for this.....

Fundamentally, I was doing a `pip install git+https://github.com/perllaghu/noteable-laf,` - but the extension neither loaded, not appeared in the list of extensions when I ran `jupyter lbaextension list`

> If you do not want to upload the artifacts from dist/ to PyPI you can still store them someplace else.

Aha.... and I see that `/dist` is in `.gitignore` .... that'll give me somewhere to explore

Thankee
